src/compiler/core.ts

```diff
@@ -242,8 +259,14 @@ namespace ts {
             const count = array.length;
             if (count > 0) {
                 let pos = 0;
-                let result = arguments.length <= 2 ? array[pos] : initial;
-                pos++;
+                let result: T | U;

```

Comment:
Assigning result to initial to start is an unnecessary extra assignment in some cases. If I change the signature of the function, I can also avoid all of the casts (though either case has no impact on actual runtime performance):

```ts
    export function reduceLeft<T, U>(array: T[], f: (a: U, x: T) => U, initial: U): U;
    export function reduceLeft<T>(array: T[], f: (a: T, x: T) => T): T;
    export function reduceLeft<T>(array: T[], f: (a: T, x: T) => T, initial?: T): T {
      ...
    }
```